rclrs/src/wait.rs

```diff
@@ -53,100 +65,115 @@ impl WaitSet {
             rcl_wait_set_init(
                 &mut rcl_wait_set as *mut _,
                 number_of_subscriptions,
-                number_of_guard_conditions,
-                number_of_timers,
-                number_of_clients,
-                number_of_services,
-                number_of_events,
-                context as *mut _,
+                0,
+                0,
+                0,
+                0,
+                0,
+                &mut *context.handle.lock() as *mut _,
                 rcutils_get_default_allocator(),
             )
             .ok()?;
             rcl_wait_set
         };
         Ok(Self {
             handle: rcl_wait_set,
+            _context_handle: context.handle.clone(),
+            subscriptions: Vec::new(),
         })
     }
 
-    /// Removes (sets to NULL) all entities in the WaitSet
+    /// Removes all entities from the wait set.
     ///
-    /// # Errors
-    /// - `RclError::InvalidArgument` if any arguments are invalid.
-    /// - `RclError::WaitSetInvalid` if the WaitSet is already zero-initialized.
-    /// - `RclError::Error` for an unspecified error
-    pub fn clear(&mut self) -> Result<(), RclReturnCode> {
-        unsafe { to_rcl_result(rcl_wait_set_clear(&mut self.handle as *mut _)) }
+    /// This effectively resets the wait set to the state it was in after being created by
+    /// [`WaitSet::new`].
+    pub fn clear(&mut self) {
+        self.subscriptions.clear();
+        // This cannot fail – the rcl_wait_set_clear function only checks that the input handle is
+        // valid, which it always is in our case. Hence, only debug_assert instead of returning
+        // Result.
+        // SAFETY: No preconditions for this function (besides passing in a valid wait set).
+        let ret = unsafe { rcl_wait_set_clear(&mut self.handle as *mut _) };
+        debug_assert_eq!(ret, 0);
     }
 
-    /// Adds a subscription to the WaitSet
+    /// Adds a subscription to the wait set.
     ///
-    /// # Errors
-    /// - `WaitSetError::DroppedSubscription` if the passed weak pointer refers to a dropped subscription
-    /// - `WaitSetError::RclError` for any `rcl` errors that occur during the process
+    /// It is possible, but not useful, to add the same subscription twice.
+    ///
+    /// This will return an error if the number of subscriptions in the wait set is larger than the
+    /// capacity set in [`WaitSet::new`].
+    ///
+    /// The same subscription must not be added to multiple wait sets, because that would make it
+    /// unsafe to simultaneously wait on those wait sets.
```

Comment:
Perhaps someday we can brainstorm some way to make sure that something is not added to multiple wait sets - but I'm not sure it's possible without significant performance overhead, and is outside the scope of this PR anyways.

Reply:
Yep.